net: mqtt-sn: Fix mqtt-sn disconnection in multiple threads.

Cherish-Gww opened the pull request:
…hreads.

memset can affect the creation of new mqtt-sn message in multiple thread enviroment.

Cherish-Gww committed:
net: mqtt-sn: Improve thread safety of publish/topic allocators

Current MQTT-SN topic and publish allocators are not thread safe.
Fix this, by using k_mem_slab instead of arrays.

Signed-off-by: Weiwei Guo <guoweiwei@syriusrobotics.com>

rlubos reviewed (commented):
It looks to me that we'd need a mutex protection for `mqtt_sn_publish pubs[]` allocation/deallocation.

Cherish-Gww commented:
> It looks to me that we'd need a mutex protection for `mqtt_sn_publish pubs[]` allocation/deallocation.

I am testing use slab instead *_get_empty, but I found that unable to send data.

rlubos reviewed (commented):
The change looks good, using k_mem_slab seems like the right choice here. I suggest though to modify the commit message a bit:
```
net: mqtt-sn: Improve thread safety of publish/topic allocators

Current MQTT-SN topic and publish allocators are not thread safe. 
Fix this, by using k_mem_slab instead of arrays.

Signed-off-by: Weiwei Guo <guoweiwei@syriusrobotics.com>
```

rlubos reviewed (dismissed):
LGTM

Cherish-Gww commented:
> LGTM

I also found a hidden bug, in function `mqtt_sn_publish`
when slab run out of buffer, `mqtt_sn_publish_create` can not create message, the mqtt_sn will be stop, because no thread will reschedule mqtt_sn process work.

carlescufi commented:
> > LGTM
> 
> I also found a hidden bug, in function `mqtt_sn_publish` when slab run out of buffer, `mqtt_sn_publish_create` can not create message, the mqtt_sn will be stop, because no thread will reschedule mqtt_sn process work.

Can you please open an issue @Cherish-Gww ?

carlescufi commented:
@Cherish-Gww please rebase

Cherish-Gww committed:
net: mqtt-sn: Active mqtt-sn process work when buffer run out

When publish buffer run out, no thread active process work.
Fix this, by reschedule process work.

Signed-off-by: Weiwei Guo <guoweiwei@syriusrobotics.com>

Cherish-Gww commented:
> 

https://github.com/zephyrproject-rtos/zephyr/pull/60728/commits/d9637f6b2cdea3a5ee03616612442ef5dd0db9dd

Cherish-Gww commented:
> 

@carlescufi 
here I fixed it: https://github.com/zephyrproject-rtos/zephyr/pull/60728/commits/d9637f6b2cdea3a5ee03616612442ef5dd0db9dd

jukkar commented on subsys/net/lib/mqtt_sn/mqtt_sn.c:
```diff
@@ -308,7 +282,7 @@ static void mqtt_sn_topic_destroy_all(struct mqtt_sn_client *client)
 		}
 
-		memset(topic, 0, sizeof(*topic));
+		k_mem_slab_free(&topics, (void **)&topic);
```
Same here

Cherish-Gww commented on subsys/net/lib/mqtt_sn/mqtt_sn.c:
```diff
@@ -126,13 +124,12 @@ static void mqtt_sn_con_init(struct mqtt_sn_confirmable *con)
 {
 	sys_slist_find_and_remove(&client->publish, &pub->next);
-	memset(pub, 0, sizeof(*pub));
+	k_mem_slab_free(&pubs, (void **)&pub);
```
Thank you, I will fix it.

jukkar commented on subsys/net/lib/mqtt_sn/mqtt_sn.c:
```diff
@@ -126,13 +124,12 @@ static void mqtt_sn_con_init(struct mqtt_sn_confirmable *con)
 {
 	sys_slist_find_and_remove(&client->publish, &pub->next);
-	memset(pub, 0, sizeof(*pub));
+	k_mem_slab_free(&pubs, (void *)&pub);
```
Changing casting does not help here, you are still passing double pointer to the free func. You need to change this to `pub` and not its address.

Cherish-Gww replied:
oh shit, I'm blind

jukkar commented on subsys/net/lib/mqtt_sn/mqtt_sn.c:
```diff
@@ -142,32 +139,21 @@ static void mqtt_sn_publish_destroy_all(struct mqtt_sn_client *client)
-		if (!mqtt_sn_pubs[i].con.in_use) {
-			return &mqtt_sn_pubs[i];
-		}
+		k_mem_slab_free(&pubs, (void *)&pub);
```
Same here.

jukkar commented on subsys/net/lib/mqtt_sn/mqtt_sn.c:
```diff
@@ -308,7 +282,7 @@ static void mqtt_sn_topic_destroy_all(struct mqtt_sn_client *client)
 		}
 
-		memset(topic, 0, sizeof(*topic));
+		k_mem_slab_free(&topics, (void *)&topic);
```
And here

Cherish-Gww commented:
@jukkar 
In the future, mqtt-sn will use net buf instead, Do you think I should be give up using slab?

jukkar commented:
> In the future, mqtt-sn will use net buf instead, Do you think I should be give up using slab?

I do not know, it is up to you to propose a PR and then community can review.

jukkar reviewed (approved):
LGTM

jukkar commented:
> How does the community run? I feel like merging patches is like holding some kind of ceremony.
> Mysterious and captivating

The community here means anyone reviewing / reading the proposed PRs. If the change/fix makes sense, people can +1 it, and if there are two acceptors the PR can be merged. It is quite simple actually :smiley:

jukkar commented:
@Cherish-Gww please rebase against latest upstream and re-submit. The culprit that caused CI to fail is now fixed in main.

Cherish-Gww commented:
> @Cherish-Gww please rebase against latest upstream and re-submit. The culprit that caused CI to fail is now fixed in main.

Wow, you have 2884 commit for zephyr, That's cool. Thank you for review my PR.

jukkar reviewed (approved):
LGTM

jukkar commented:
> Thank you for review my PR.

I am one of the maintainers in networking area so I am reviewing the PRs all the time. You, as a member of the community (because you are sending PRs), can also review other people PRs if you want.

ssharks commented:
@Cherish-Gww Working on such a big project as Zephyr requires some discipline. Especially in the first few PR's that can be a little frustrating due to the steep learning curve with all rules, process and CI/CD and review. All is in place to minimize the risk of an accidental error affecting 100's of users.
I'm a rather new reviewer and as I have not enough experience on this particular area, I'll leave this one for another reviewer. He will come to it, but it will take a while.
